File: src/Will_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), count_(0)
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>
        ("number", 10, std::bind(&NumberCounterNode::callbackNumber, this, std::placeholders::_1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter", std::bind(&NumberCounterNode::callbackReset, this, _1,_2));
    }
private:
    void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        count_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "%d", count_);
        auto msg2 = example_interfaces::msg::Int64();
        msg2.data = count_;
        publisher_->publish(msg2);
    }

    void callbackReset(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                       const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data)
        {
            count_ = 0;
            response->success = true;
            response->message = std::string("reset !");  
        }  
        else
        {
            response->success = false;
            response->message = std::string("what the hell...");  
        }
        RCLCPP_INFO(this->get_logger(), response->message.c_str());
    }

    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
    int count_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3 
import rclpy 
from rclpy.node import Node 
from example_interfaces.msg import Int64, String
from example_interfaces.srv import SetBool

class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.subscriber_ = self.create_subscription(Int64, "number", self.callback_number, 10)
        self.get_logger().info("Number counter is up and running!")
        self.publishers_ = self.create_publisher(Int64, "number_count", 10)
        self.counter_ = 0
        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
    
    def callback_number(self, msg: Int64):
        self.get_logger().info("I received: " + str(msg.data) + " ! ")
        self.counter_ += 1
        msg_count = Int64()
        msg_count.data = msg.data * self.counter_
        self.publishers_.publish(msg_count)

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data == True:
            self.counter_ = 0
            response.success = True
            response.message = "Counter reset to 0"
        else:
            response.success = False
            response.message = "No reset!"
        return response

def main(args=None): 
    rclpy.init(args=args) 
    node = NumberCounterNode()  
    rclpy.spin(node)    
    rclpy.shutdown() 

if __name__ == "__main__":  
    main() 